## Hands On - Trabalho final 
### Netflix Movies and TV Shows

#### Grupo de trabalho
1. João Freitas,
2. João Kruschewsky,
3. Vinícius Matos,
4. Vito Magarão


#### Objetivo

Sobre este conjunto de dados: Netflix é uma das plataformas de streaming de mídia e vídeo mais populares. Eles têm mais de 8.000 filmes ou programas de TV disponíveis em sua plataforma. Em meados de 2021, eles têm mais de 200 milhões de assinantes em todo o mundo. Este conjunto de dados tabulares consiste em listas de todos os filmes e programas de TV disponíveis no Netflix, juntamente com detalhes como elenco, diretores, classificações, ano de lançamento, duração, etc.

1. Entender qual conteúdo está disponível em diferentes países.
2. Identificando conteúdo semelhante combinando recursos baseados em texto.
3. Análise de rede de atores/diretores e encontre insights interessantes.
4. A Netflix tem mais foco em programas de TV do que em filmes nos últimos anos?

In [70]:
import pandas as pd
import matplotlib as plt

In [71]:
df = pd.read_csv("https://raw.githubusercontent.com/MasterExecuson/Data-Science-Analytics/main/Data%20Analytics%20Hands%20On/Hands-on%20-%20Final%20Project/Netflix%20Movies%20and%20TV%20Shows%20dataset/netflix_titles.csv")

In [72]:
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [73]:
df.tail()

show_id     type        title         director  \
8802   s8803    Movie       Zodiac    David Fincher   
8803   s8804  TV Show  Zombie Dumb              NaN   
8804   s8805    Movie   Zombieland  Ruben Fleischer   
8805   s8806    Movie         Zoom     Peter Hewitt   
8806   s8807    Movie       Zubaan      Mozez Singh   

                                                   cast        country  \
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

             date_added  release_year rating   duration  \
8802  November 20, 2019          2007      R    158 min   
8803       July 1, 2019          2018  TV-Y7  2 Seasons   
8804   November 1, 2019          2009      R     88 min   
8805   January 11, 2020          2006     PG     88 min   
8806      March 2, 2019          2015  TV-14    111 min   

                                           listed_in  \
8802                  Cult Movies, Dramas, Thrillers   
8803          Kids' TV, Korean TV Shows, TV Comedies   
8804                         Comedies, Horror Movies   
8805              Children & Family Movies, Comedies   
8806  Dramas, International Movies, Music & Musicals   

                                            description  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...

In [74]:
df.head()['description'][0]

'As her father nears the end of his life, filmmaker Kirsten Johnson stages his death in inventive and comical ways to help them both face the inevitable.'

In [75]:
for column in df.columns:
    print(df[column].unique())

['s1' 's2' 's3' ... 's8805' 's8806' 's8807']
['Movie' 'TV Show']
['Dick Johnson Is Dead' 'Blood & Water' 'Ganglands' ... 'Zombieland'
 'Zoom' 'Zubaan']
['Kirsten Johnson' nan 'Julien Leclercq' ... 'Majid Al Ansari'
 'Peter Hewitt' 'Mozez Singh']
[nan
 'Ama Qamata, Khosi Ngema, Gail Mabalane, Thabang Molaba, Dillon Windvogel, Natasha Thahane, Arno Greeff, Xolile Tshabalala, Getmore Sithole, Cindy Mahlangu, Ryle De Morny, Greteli Fincham, Sello Maake Ka-Ncube, Odwa Gwanya, Mekaila Mathys, Sandi Schultz, Duane Williams, Shamilla Miller, Patrick Mofokeng'
 'Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabiha Akkari, Sofia Lesaffre, Salim Kechiouche, Noureddine Farihi, Geert Van Rampelberg, Bakary Diombera'
 ...
 'Jesse Eisenberg, Woody Harrelson, Emma Stone, Abigail Breslin, Amber Heard, Bill Murray, Derek Graf'
 'Tim Allen, Courteney Cox, Chevy Chase, Kate Mara, Ryan Newman, Michael Cassidy, Spencer Breslin, Rip Torn, Kevin Zegers'
 'Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanana, Manish Ch

In [76]:
def get_unique_chars(df): 
    text = ''.join(df['comentario'].astype(str).values.flatten())
    unique_chars = list(set(text))
    return unique_chars


# Limpeza de colunas do dataframe para o primeiro parse 
def cleaning_df (df,column_name):
    df_clean = df.copy()
    exclude_chars = ["’", "—", "–",'™']
    for char in exclude_chars:
        # Essa expressão regular remove todos os caracteres que não estão no alfabeto ou que possuem comprimento menor que 2. Ela também removerá "s" e sinais de pontuação.
        df_clean[column_name] = df_clean[column_name].replace(to_replace=r'[^a-zA-Z]|(\b\w{1}\b)', value=' ', regex=True)
    df_clean[column_name] = df_clean[column_name].replace(regex=r' +', value=' ') # remove quaisquer espaços duplicados ou triplicados que foram gerados como resultado da primeira linha.
    df_clean[column_name] = [content.translate(str.maketrans('','', string.punctuation)) for content in df_clean[column_name]]
    df_clean[column_name] = [content.translate(str.maketrans('','', string.digits)) for content in df_clean[column_name]]
    df_clean[column_name] = df_clean[column_name].str.strip()
    return(df_clean)

In [77]:
no_comentario_unique = len(df['description'].unique())
print(f'O dataset apresenta {df.shape[0]} entradas e {df.shape[1]} atributos.')
print(f'A coluna "description", que é a que contem os comentários, possui {no_comentario_unique} entradas únicas.')


O dataset apresenta 8807 entradas e 12 atributos.
A coluna "description", que é a que contem os comentários, possui 8775 entradas únicas.


In [78]:
reated_lines = df.shape[0]-len(df['description'].unique())
print(f'Podemos ver que possuimos {reated_lines} comentários repetidos.')

Podemos ver que possuimos 32 comentários repetidos.


In [79]:
print(f'O dataset apresenta {df.duplicated().sum()} linhas duplicadas.')
print("Podemos ver que apesar de possuirmos descrições iguais, não há linhas idênticas.")

O dataset apresenta 0 linhas duplicadas.
Podemos ver que apesar de possuirmos descrições iguais, não há linhas idênticas.


In [80]:
print("Podemos notar que através da análise que temos viversas linhas repetidas poucas vezes.\
       Se tratam de 27 comentários diferentes com repetições variadas de 4 a 2 instâncias.")
df_grouped = df[['show_id','description']].groupby(by='description').count().sort_values(by='show_id',ascending=False)
repeated_comments = df_grouped[df_grouped['show_id'] > 1]
print(repeated_comments.count())

Podemos notar que através da análise que temos viversas linhas repetidas poucas vezes.       Se tratam de 27 comentários diferentes com repetições variadas de 4 a 2 instâncias.
show_id    27
dtype: int64


In [69]:
repeated_df = df[df['description'].isin(repeated_comments.index)]

In [82]:
print(repeated_df.count())
print("Podemos ver que possuimos 59 shows que possuem a pelo menos uma discrição igual a outro. Essa informação é corroborada pelo title, que possui a mesma cardinalidade.")

show_id         59
type            59
title           59
director        46
cast            52
country         39
date_added      59
release_year    59
rating          59
duration        59
listed_in       59
description     59
dtype: int64
Podemos ver que possuimos 59 shows que possuem a pelo menos uma discrição igual a outro. Essa informação é corroborada pelo title, que possui a mesma cardinalidade.


In [90]:
repeated_df[['title','description']].sort_values(by='description')

title  \
78                                       Tughlaq Durbar   
79                              Tughlaq Durbar (Telugu)   
7022                               Hum Saath-Saath Hain   
2969                              Together For Eternity   
3492                               Oh! Baby (Malayalam)   
3516                                           Oh! Baby   
3493                                   Oh! Baby (Tamil)   
8051                            Solo: A Star Wars Story   
8052          Solo: A Star Wars Story (Spanish Version)   
5965                                             22-Jul   
4522                                            22 July   
2840                   Angu Vaikuntapurathu (Malayalam)   
2873                            Ala Vaikunthapurramuloo   
3933                             Petta (Telugu Version)   
3939                                              Petta   
3932               Sarvam Thaala Mayam (Telugu Version)   
4061                Sarvam Thaala Mayam (Tamil Version)   
1652                                       Andhaghaaram   
1653                                        Andhakaaram   
3996                                         February 9   
5964                                             Feb-09   
3569                          Game Over (Tamil Version)   
3568                          Game Over (Hindi Version)   
1986                                   Nee Enge En Anbe   
1982                                           Anaamika   
1270                           Sin senos sí hay paraíso   
8022                           Sin Senos sí Hay Paraíso   
852                                   99 Songs (Telugu)   
851                                    99 Songs (Tamil)   
850                                            99 Songs   
303                                Esperando la carroza   
6705                               Esperando La Carroza   
8173                                          Thackeray   
2336                                  Thackeray (Hindi)   
7559                    Naruto Shippuden : Blood Prison   
56             Naruto Shippuden the Movie: Blood Prison   
2335                                     Seven (Telugu)   
2334                                      Seven (Tamil)   
6024                                          7 (Seven)   
5966                                             15-Aug   
3962                                          15 August   
236                                             Boomika   
239                                    Boomika (Telugu)   
238                                 Boomika (Malayalam)   
237                                     Boomika (Hindi)   
6529                                      Consequences    
3371                                       Consequences   
52    InuYasha the Movie 3: Swords of an Honorable R...   
7089  Inuyasha the Movie - La spada del dominatore d...   
2664      ChuChu TV Nursery Rhymes & Kids Songs (Hindi)   
3393  ChuChu TV Kids Songs, Learning Videos & Bedtim...   
7649                                       Ollie & Moon   
3028                              The Ollie & Moon Show   
8360                                  The Incredibles 2   
7067                    Incredibles 2 (Spanish Version)   
6449                                    Chashme Buddoor   
2270                                    Chashme Baddoor   
7090  InuYasha: The Movie 2: The Castle Beyond the L...   
51    InuYasha the Movie 2: The Castle Beyond the Lo...   

                                            description  
78    A budding politician has devious plans to rise...  
79    A budding politician has devious plans to rise...  
7022  A scheming matriarch plots to cut off her disa...  
2969  A scheming matriarch plots to cut off her disa...  
3492  A surly septuagenarian gets another chance at ...  
3516  A surly septuagenarian gets another chance at ...  
3493  A surly septuagenarian gets another chance at ...  
8051  A young Han Solo tries to settle an old

Podemos ver através da análise acima que os títulos com descrição repetida realmente é apenas uma outra versão do título original.


In [91]:
repeated_df.sort_values(by='description')

show_id     type                                              title  \
78       s79    Movie                                     Tughlaq Durbar   
79       s80    Movie                            Tughlaq Durbar (Telugu)   
7022   s7023    Movie                               Hum Saath-Saath Hain   
2969   s2970    Movie                              Together For Eternity   
3492   s3493    Movie                               Oh! Baby (Malayalam)   
3516   s3517    Movie                                           Oh! Baby   
3493   s3494    Movie                                   Oh! Baby (Tamil)   
8051   s8052    Movie                            Solo: A Star Wars Story   
8052   s8053    Movie          Solo: A Star Wars Story (Spanish Version)   
5965   s5966    Movie                                             22-Jul   
4522   s4523    Movie                                            22 July   
2840   s2841    Movie                   Angu Vaikuntapurathu (Malayalam)   
2873   s2874    Movie                            Ala Vaikunthapurramuloo   
3933   s3934    Movie                             Petta (Telugu Version)   
3939   s3940    Movie                                              Petta   
3932   s3933    Movie               Sarvam Thaala Mayam (Telugu Version)   
4061   s4062    Movie                Sarvam Thaala Mayam (Tamil Version)   
1652   s1653    Movie                                       Andhaghaaram   
1653   s1654    Movie                                        Andhakaaram   
3996   s3997  TV Show                                         February 9   
5964   s5965  TV Show                                             Feb-09   
3569   s3570    Movie                          Game Over (Tamil Version)   
3568   s3569    Movie                          Game Over (Hindi Version)   
1986   s1987    Movie                                   Nee Enge En Anbe   
1982   s1983    Movie                                           Anaamika   
1270   s1271  TV Show                           Sin senos sí hay paraíso   
8022   s8023  TV Show                           Sin Senos sí Hay Paraíso   
852     s853    Movie                                  99 Songs (Telugu)   
851     s852    Movie                                   99 Songs (Tamil)   
850     s851    Movie                                           99 Songs   
303     s304    Movie                               Esperando la carroza   
6705   s6706    Movie                               Esperando La Carroza   
8173   s8174  TV Show                                          Thackeray   
2336   s2337    Movie                                  Thackeray (Hindi)   
7559   s7560    Movie                    Naruto Shippuden : Blood Prison   
56       s57    Movie           Naruto Shippuden the Movie: Blood Prison   
2335   s2336    Movie                                     Seven (Telugu)   
2334   s2335    Movie                                      Seven (Tamil)   
6024   s6025  TV Show                                          7 (Seven)   
5966   s5967    Movie                                             15-Aug   
3962   s3963    Movie                                          15 August   
236     s237    Movie                                            Boomika   
239     s240    Movie                                   Boomika (Telugu)   
238     s239    Movie                                Boomika (Malayalam)   
237     s238    Movie                                    Boomika (Hindi)   
6529   s6530    Movie                                      Consequences    
3371   s3372    Movie                                       Consequences   
52       s53    Movie  InuYasha the Movie 3: Swords of an Honorable R...   
7089   s7090    Movie  Inuyasha the Movie - La spada del dominatore d...   
2664   s2665  TV Show      ChuChu TV Nursery Rhymes & Kids Songs (Hindi)   
3393   s3394  TV Show  ChuChu TV Kids Songs, Learning Videos & Bedtim...   
7649   s7650  TV Show                                

Podemos ver através da análise acima que a maior parte dos dados repetidos possuem a mesma informação, porém alguns títulos possuem informações divergentes como:

- 851	s852	Movie	99 Songs (Tamil)	NaN	NaN	NaN	May 21, 2021	2021	TV-14	131 min	Dramas, International Movies, Music & Musicals	Challenged to compose 100 songs before he can ...
- 850	s851	Movie	99 Songs	Vishwesh Krishnamoorthy	Ehan Bhat, Edilsy Vargas, Manisha Koirala, Lis...	India	May 21, 2021	